### PR TITLE
API Ownership API Implementation

### PR DESCRIPTION
**Pull request by tractorcow**
Implementation of first feature as a part of Neo-Versioning RFC

https://github.com/silverstripe/silverstripe-framework/issues/4932

This change required a few other changes which have been broken out, and can be merged independently:

- https://github.com/silverstripe/silverstripe-framework/pull/5094
- https://github.com/silverstripe/silverstripe-framework/pull/5081

CMS pull request link:
- https://github.com/silverstripe/silverstripe-cms/pull/1401

### CODE REVIEW

**Comment by tractorcow** [comment]
Attempts to break up this PR into multiple commits seemed to have wrecked the tests... squashing for now.

**Comment by tractorcow** [comment]
Doh, i used php 5.6 only code. Will fix to work with 5.5. :)

File: docs/en/02_Developer_Guides/00_Model/10_Versioning.md [hunk 1]
```diff
@@ -129,6 +129,77 @@ is initialized. But it can also be set and reset temporarily to force a specific
 	$obj = MyRecord::getComplexObjectRetrieval(); // returns 'Stage' records
 	Versioned::set_reading_mode($origMode); // reset current mode
 
+### File ownership
+
+Typically when publishing versioned dataobjects, it is necessary to ensure that some linked components
+are published along with it. Unless this is done site front-end content can appear incorrectly published.
```
Comment (chillu) [inline docs/en/02_Developer_Guides/00_Model/10_Versioning.md]: Add comma: "Unless this is done, site ..." - a bit hard to read currently.

File: docs/en/02_Developer_Guides/00_Model/10_Versioning.md [hunk 2]
```diff
@@ -129,6 +129,77 @@ is initialized. But it can also be set and reset temporarily to force a specific
...
+	:::php
+	class MyPage extends Page {
+		private static $has_many = array(
+			'Banners' => 'Banner'
+		);
+		private static $owns = array(
+			'Banners'
+		);
+		
+		public function onAfterVersionedPublish($fromStage, $toStage, $createNewVersion) {
```
Comment (chillu) [inline docs/en/02_Developer_Guides/00_Model/10_Versioning.md]: Assuming the CMS will take care of this automatically through changesets, is it helpful to add a custom "auto-publish" implementation in this high level documentation? It gives the impression that you need to write custom hooks for this feature to work.

File: model/versioning/Versioned.php [hunk 3]
```diff
@@ -161,6 +162,34 @@ class Versioned extends DataExtension implements TemplateGlobalProvider {
...
+	 */
+	private static $owns = array();
+
+	/**
+	 * Opposing relationship to owns config; Represents the objects which
+	 * own the current object.
+	 *
+	 * @var array
+	 */
+	private static $owned_by = array();
```
Comment (chillu) [inline model/versioning/Versioned.php]: Document if `$owned_by` is optional or not

File: model/versioning/Versioned.php [hunk 4]
```diff
@@ -780,6 +805,95 @@ public function onAfterSkippedWrite() {
...
+
+		// Skip search for unsaved records
+		if(!$this->owner->isInDB()) {
+			return $list;
+		}
+
+		$relationships = $this->owner->config()->{$source};
+		foreach($relationships as $relationship) {
+			// Warn if invalid config
+			if(!$this->owner->hasMethod($relationship)) {
```
Comment (chillu) [inline model/versioning/Versioned.php]: Does that mean you can declare ownership on arbitrary getters? If it's not the case, we should also check if `$relationship` is actually a DO relationship

File: model/versioning/Versioned.php [hunk 5]
```diff
@@ -1174,7 +1288,7 @@ public function publish($fromStage, $toStage, $createNewVersion = false) {
 		$conn = DB::get_conn();
 		if(method_exists($conn, 'allowPrimaryKeyEditing')) {
 			$conn->allowPrimaryKeyEditing($baseClass, true);
-			$from->write();
+		$from->write();
```
Comment (chillu) [inline model/versioning/Versioned.php]: Wrong indentation?

File: model/versioning/Versioned.php [hunk 6]
```diff
@@ -780,6 +805,95 @@ public function onAfterSkippedWrite() {
...
+		if (!$list) {
+			$list = new ArrayList();
+		}
+
+		// Skip search for unsaved records
+		if(!$this->owner->isInDB()) {
+			return $list;
+		}
+
+		$relationships = $this->owner->config()->{$source};
```
Comment (chillu) [inline model/versioning/Versioned.php]: The RFC states:

> Ownership, unlike other relationship types, is only declared on the owner object via the configuration system. I.e. owns not owned_by is set. Owned by is implicit by any many_many, has_many, blongs_to, has_one, or belongs_many_many where the other side declared this object as owned

Does that mean you need to do more work here for `$source == 'owned_by'`?

File: docs/en/02_Developer_Guides/00_Model/10_Versioning.md (hunk 2 above)
Comment (tractorcow) [inline docs/en/02_Developer_Guides/00_Model/10_Versioning.md]: I'll remove the "temp code" part, and assume we'll make it work during changesets. ;)

File: model/versioning/Versioned.php (hunk 3 above)
Comment (tractorcow) [inline model/versioning/Versioned.php]: Docs say

```
For instance, in order to specify this dependency, you must apply `owns` and `owned_by` config
on a relationship.
```

File: model/versioning/Versioned.php (hunk 4 above)
Comment (tractorcow) [inline model/versioning/Versioned.php]: It's not an explicit requirement, but it does work. ;)

File: model/versioning/Versioned.php (hunk 6 above)
Comment (tractorcow) [inline model/versioning/Versioned.php]: I changed my mind, for performance reasons.

**Comment by tractorcow** [comment]
Annoyingly, testing the failing test locally on Postgres reveals no errors at all. =/

**Comment by tractorcow** [comment]
Tests green! Just one last push to squash it all up.

**Comment by tractorcow** [comment]
:shipit:

**Comment by chillu** [comment]
Sweet, happy with this now. I've split out the validation into an enhancement story in our team backlog, with the following content:

Acceptance Criteria:
 * As a developer, I'll find out quickly if I've misconfigured the ownership API on my DataObjects in order to avoid frustration on why objects aren't included in publication
 * As a developer, I can develop with SilverStripe as quickly as I've been used to in 3.x

Description:
At the moment, both sides need to be defined ($owns and $owned_by). This was a simplification to get the initial implementation over the line. Virtually every dev will need to manually create these relationships for potentially dozens of custom DataObject definitions when upgrading to 4.x. This means we can't have any misconfigurations fail silently (exclude objects from publication). A few options are possible:

- Don't require the reflexive relationship, infer it on each findOwned() call and cache
- Enforce checks on dev/build
- Enforce checks on a new dev/validate endpoint
- Enforce checks on each findOwned() call and cache

**Comment by chillu** [comment]
Can we get another +1 from @silverstripe/core-team please?

**Comment by tractorcow** [comment]
Can someone merge this, so that I can raise the PR for the next stage? 

I'm about to implement triggered versioning and it depends on this. :)

@dhensby ? @hafriedlander ?

**Comment by chillu** [comment]
Merging, given enough opportunity for feedback now (incl. more than a month on the underlying RFC), and it's starting to block our work

**Pull request merged**